# 3. Practica sobre procesamiento y exploración de datos

    1. Importar los datos liga española de primera división posteriores al break Covid-19.(/datos/all-euro-data-2020-2021.xlsx, hoja 'SP1'.
    
    2. Limpiar todos los valores nulos imputando para cada uno de ellos la mediana de su columna.
    
    3. Calcular con funciones de dataframe la media y desviación estándar de las columnas ['HTHG','HTAG','FTHG','FTAG'] que corresponden a los goles en el descanso del equipo de casa y visitante, y al final del partido del equipo de casa y visitante, respectivamente.
    
    4. Calcular qué pareja de variables de las mencionadas tienen mayor correlación.
    
    5. Calcular la matriz de covarianza sobre las anteriores variables y hacer un ranking de mayor a menor varianza.
    
    6. Discutir cómo varía cada variable con respecto a las otras usando los valores de la matriz de covarianza.
    
    7. Hacer un label encoding con los campos 'Div' (liga), 'HomeTeam' y 'AwayTeam'.
    
    8. Normalizar las variables ['HTHG','HTAG','FTHG','FTAG'] y volver a calcular la matriz de covarianza. Discusión sobre la matriz de covarianza normalizada.
    
    9. Escribir una función que tome dos valores de score (goles home team, goles away team) y devuelva 'H' si gana el de casa, 'A' el de fuera o 'D' en caso de empate.
    
    10. Ejecutar la función anterior usando la función .apply(lambda) sobre toda la matriz con variables ['FTHG','FTAG'] escribiendo el resultado en una nueva columna llamada 'resultado'.

## 1.-

In [1]:
import pandas as pd
import numpy as np

data=pd.read_excel('datos/all-euro-data-2020-2021.xlsx',sheet_name='SP1')

## 2.-

In [3]:
print(data.head(100))
data=data.fillna(data.median())
data.head(100)

    Div       Date      Time     HomeTeam     AwayTeam  FTHG  FTAG FTR  HTHG  \
0   SP1 2020-09-12  15:00:00        Eibar        Celta     0     0   D     0   
1   SP1 2020-09-12  17:30:00      Granada   Ath Bilbao     2     0   H     0   
2   SP1 2020-09-12  20:00:00        Cadiz      Osasuna     0     2   A     0   
3   SP1 2020-09-13  13:00:00       Alaves        Betis     0     1   A     0   
4   SP1 2020-09-13  15:00:00   Valladolid     Sociedad     1     1   D     1   
5   SP1 2020-09-13  17:30:00   Villarreal       Huesca     1     1   D     0   
6   SP1 2020-09-13  20:00:00     Valencia      Levante     4     2   H     2   
7   SP1 2020-09-19  15:00:00   Villarreal        Eibar     2     1   H     0   
8   SP1 2020-09-19  17:30:00       Getafe      Osasuna     1     0   H     0   
9   SP1 2020-09-19  20:00:00        Celta     Valencia     2     1   H     1   
10  SP1 2020-09-20  15:00:00       Huesca        Cadiz     0     2   A     0   
11  SP1 2020-09-20  17:30:00        Beti

Div       Date      Time     HomeTeam     AwayTeam  FTHG  FTAG FTR  HTHG  \
0   SP1 2020-09-12  15:00:00        Eibar        Celta     0     0   D     0   
1   SP1 2020-09-12  17:30:00      Granada   Ath Bilbao     2     0   H     0   
2   SP1 2020-09-12  20:00:00        Cadiz      Osasuna     0     2   A     0   
3   SP1 2020-09-13  13:00:00       Alaves        Betis     0     1   A     0   
4   SP1 2020-09-13  15:00:00   Valladolid     Sociedad     1     1   D     1   
5   SP1 2020-09-13  17:30:00   Villarreal       Huesca     1     1   D     0   
6   SP1 2020-09-13  20:00:00     Valencia      Levante     4     2   H     2   
7   SP1 2020-09-19  15:00:00   Villarreal        Eibar     2     1   H     0   
8   SP1 2020-09-19  17:30:00       Getafe      Osasuna     1     0   H     0   
9   SP1 2020-09-19  20:00:00        Celta     Valencia     2     1   H     1   
10  SP1 2020-09-20  15:00:00       Huesca        Cadiz     0     2   A     0   
11  SP1 2020-09-20  17:30:00        Betis   Valladolid     2     0   H     2   
12  SP1 2020-09-20  17:30:00      Granada       Alaves     2     1   H     1   
13  SP1 2020-09-20  20:00:00     Sociedad  Real Madrid     0     0   D     0   
14  SP1 2020-09-26  12:00:00       Alaves       Getafe     0     0   D     0   
15  SP1 2020-09-26  15:00:00     Valencia       Huesca     1     1   D     1   
16  SP1 2020-09-26  17:30:00        Elche     Sociedad     0     3   A     0   
17  SP1 2020-09-26  20:00:00        Betis  Real Madrid     2     3   A     2   
18  SP1 2020-09-27  11:00:00      Osasuna      Levante     1     3   A     1   
19  SP1 2020-09-27  13:00:00        Eibar   Ath Bilbao     1     2   A     0   
20  SP1 2020-09-27  15:00:00   Ath Madrid      Granada     6     1   H     1   
21  SP1 2020-09-27  17:30:00        Cadiz      Sevilla     1     3   A     0   
22  SP1 2020-09-27  17:30:00   Valladolid        Celta     1     1   D     0   
23  SP1 2020-09-27  20:00:00    Barcelona   Villarreal     4     0   H     4   
24  SP1 2020-09-29  18:00:00     Sociedad     Valencia     0     1   A     0   
25  SP1 2020-09-29  20:30:00       Getafe        Betis     3     0   H     3   
26  SP1 2020-09-30  18:00:00       Huesca   Ath Madrid     0     0   D     0   
27  SP1 2020-09-30  18:00:00   Villarreal       Alaves     3     1   H     2   
28  SP1 2020-09-30  20:30:00        Eibar        Elche     0     1   A     0   
29  SP1 2020-09-30  20:30:00  Real Madrid   Valladolid     1     0   H     0   
..  ...        ...       ...          ...          ...   ...   ...  ..   ...   
70  SP1 2020-11-01  17:30:00      Granada      Levante     1     1   D     1   
71  SP1 2020-11-01  20:00:00     Valencia       Getafe     2     2   D     1   
72  SP1 2020-11-02  20:00:00   Villarreal   Valladolid     2     0   H     2   
73  SP1 2020-11-06  20:00:00        Elche        Celta     1     1   D     1   
74  SP1 2020-11-07  13:00:00       Huesca        Eibar     1     1   D     0   
75  SP1 2020-11-07  15:15:00    Barcelona        Betis     5     2   H     1   
76  SP1 2020-11-07  17:30:00      Sevilla      Osasuna     1     0   H     0   
77  SP1 2020-11-07  20:00:00   Ath Madrid        Cadiz     4     0   H     2   
78  SP1 2020-11-08  13:00:00       Getafe   Villarreal     1     3   A     1   
79  SP1 2020-11-08  15:15:00     Sociedad      Granada     2     0   H     2   
80  SP1 2020-11-08  17:30:00      Levante       Alaves     1     1   D     0   
81  SP1 2020-11-08  17:30:00   Valladolid   Ath Bilbao     2     1   H     1   
82  SP1 2020-11-08  20:00:00     Valencia  Real Madrid     4     1   H     2   
83  SP1 2020-11-20  20:00:00      Osasuna       Huesca     1     1   D     0   
84  SP1 2020-11-21  13:00:00      Levante        Elche     1     1   D     1   
85  SP1 2020-11-21  15:15:00   Villarreal  Real Madrid     1     1   D     0   
86  SP1 2020-11-21  17:30:00      Sevilla        Celta     4     2   H     2   
87  SP1 2020-11-21  20:00:00   Ath Madrid    Barcelona     1     0   H     1   
88  

En este caso los **datos venian bastante limpios** y no hay valores nulos practicamente.

## 3.-

In [10]:
feats=['HTHG','HTAG','FTHG','FTAG']
print('Valores medios:\n',data[feats].mean())
print()
print('Valores medios:\n',data[feats].std())

Valores medios:
 HTHG    0.600000
HTAG    0.534483
FTHG    1.337931
FTAG    1.127586
dtype: float64

Valores medios:
 HTHG    0.779495
HTAG    0.725595
FTHG    1.174826
FTAG    1.046132
dtype: float64


## 4.-

In [11]:
data[feats].corr()

HTHG      HTAG      FTHG      FTAG
HTHG  1.000000 -0.018353  0.677100 -0.056011
HTAG -0.018353  1.000000 -0.025895  0.716705
FTHG  0.677100 -0.025895  1.000000 -0.085880
FTAG -0.056011  0.716705 -0.085880  1.000000

Las **parejas más correlacionadas** (linealmente y positivamente) son: **(FTAG,HTAG) y (FTHG,HTHG)**:

 - El número de **goles** (tanto para el equipo de casa como para el visitante) marcados **en el descanso** están fuertemente (y positivamente) **correlacionados con el total de goles marcados al final** del partido (es **sentido común confirmado cuantitativamente**).
 
 - Los **goles marcados por los dos equipos (home/away)** están **correlados negativamente** (cuando una cantidad crece la otra decrece).

## 5.-

In [37]:
covarianza=data[feats].cov()
print(covarianza)
print()
varianza=[]
for feat in feats:
    for indice in covarianza.index.values:
        if feat==indice:
            varianza.append(covarianza[feat].loc[indice])
    
varianza

          HTHG      HTAG      FTHG      FTAG
HTHG  0.607612 -0.010381  0.620069 -0.045675
HTAG -0.010381  0.526488 -0.022074  0.544028
FTHG  0.620069 -0.022074  1.380217 -0.105548
FTAG -0.045675  0.544028 -0.105548  1.094392



[0.6076124567474044,
 0.5264884858608754,
 1.3802171578570575,
 1.0943920773177416]

In [38]:
varianza=np.array(varianza)
print(varianza)
varianza=np.sort(varianza)
varianza

[0.60761246 0.52648849 1.38021716 1.09439208]


array([0.52648849, 0.60761246, 1.09439208, 1.38021716])

**'FTHG' y 'FTAG' son las variables que APARENTEMENTE tienen mayor variabilidad (en ese orden)**. **Más tarde se verá lo que ocurre con estas magnitudes cuando las variables estén normalizadas.**

## 6.-

In [39]:
covarianza

HTHG      HTAG      FTHG      FTAG
HTHG  0.607612 -0.010381  0.620069 -0.045675
HTAG -0.010381  0.526488 -0.022074  0.544028
FTHG  0.620069 -0.022074  1.380217 -0.105548
FTAG -0.045675  0.544028 -0.105548  1.094392

   - Los **valores fuera de la diagonal** indican **como varía una variable cuando crece su pareja** en la matriz.

   - **Valores positivos** indican que **ambas variables crecen o decrecen a la vez**.
    
   - **Valores negativos** indican que ambas variables **se mueven en direcciones contrarias** (cuando una crece la otra decrece)
    
   - Es importante tener en cuenta que las **variables no están normalizadas** por lo que en general los **valores absolutos** para distintas parejas **no son en general comparables** con otras parejas.
   
   - En este caso concreto dado que **todas las variables estudiadas se refieren a goles** marcados se puede considerar que **hay normalización**. En el **caso general habría que normalizar** las variables **si se quisiera comparar valores absolutos** entre distintas parejas.

## 7.-

In [55]:
from sklearn.preprocessing import LabelEncoder
data=pd.read_excel('datos/all-euro-data-2020-2021.xlsx',sheet_name='SP1')
feats=['Div', 'HomeTeam' , 'AwayTeam']
encoder_liga=LabelEncoder()
encoder_teams=LabelEncoder()
data['Div']=encoder_liga.fit_transform(data['Div'].values)
teams=np.concatenate([data['HomeTeam'].values,data['AwayTeam'].values]) #Juntamos todos los equipos

encoder_teams.fit(teams)
data['HomeTeam']=encoder_teams.transform(data['HomeTeam'])
data['AwayTeam']=encoder_teams.transform(data['AwayTeam'])
print(data[feats].head())

   Div  HomeTeam  AwayTeam
0    0         7         6
1    0        10         1
2    0         5        13
3    0         0         4
4    0        18        16


## 8.-

In [58]:
from sklearn.preprocessing import StandardScaler
feats=['HTHG','HTAG','FTHG','FTAG']
scaler=StandardScaler()
data[feats]=scaler.fit_transform(data[feats])
print(data[feats].head())
print()
data[feats].cov()


      HTHG      HTAG      FTHG      FTAG
0 -0.77106 -0.737886 -1.140802 -1.079726
1 -0.77106 -0.737886  0.564520 -1.079726
2 -0.77106  0.642675 -1.140802  0.835384
3 -0.77106 -0.737886 -1.140802 -0.122171
4  0.51404 -0.737886 -0.288141 -0.122171



HTHG      HTAG      FTHG      FTAG
HTHG  1.003460 -0.018417  0.679443 -0.056205
HTAG -0.018417  1.003460 -0.025984  0.719185
FTHG  0.679443 -0.025984  1.003460 -0.086177
FTAG -0.056205  0.719185 -0.086177  1.003460

Esta es la matriz de **covarianza antes de normalizar**:

HTHG	HTAG	FTHG	FTAG
HTHG	0.607612	-0.010381	0.620069	-0.045675
HTAG	-0.010381	0.526488	-0.022074	0.544028
FTHG	0.620069	-0.022074	1.380217	-0.105548
FTAG	-0.045675	0.544028	-0.105548	1.094392

   - Los **signos no cambian**. Las **correlaciones, de forma cualitativa, no han cambiado.
    
   - **Cambian las magnitudes**, y **ahora sí son comparables** al ser variables normalizadas.
   
   - **La varianza es la misma para todas las variable**.

## 9.-

In [65]:
def myfunction(goles_h,goles_a):
    if goles_h>goles_a:
        result='H'
    elif goles_h<goles_a:
        result='A'
    else:
        result='D'
    return result

## 10.-

In [77]:
data=pd.read_excel('datos/all-euro-data-2020-2021.xlsx',sheet_name='SP1')

In [78]:
feats=['FTHG','FTAG'] 
data['result']=data[feats].apply(lambda x: myfunction(*x),axis=1)
feats=feats+['result']
data[feats].head()



FTHG  FTAG result
0     0     0      D
1     2     0      H
2     0     2      A
3     0     1      A
4     1     1      D